coderd/wsbuilder/wsbuilder.go

```diff
@@ -0,0 +1,741 @@
+	// nolint: revive
+	b.state = stateTarget{orphan: true}
+	return b
+}
+
+func (b Builder) LogLevel(l string) Builder {
+	// nolint: revive
+	b.logLevel = l
+	return b
+}
+
+func (b Builder) Initiator(u uuid.UUID) Builder {
+	// nolint: revive
+	b.initiator = u
+	return b
+}
+
+func (b Builder) Reason(r database.BuildReason) Builder {
+	// nolint: revive
+	b.reason = r
+	return b
+}
+
+func (b Builder) LegacyParameterValues(p []codersdk.CreateParameterRequest) Builder {
+	// nolint: revive
+	b.legacyParameterValues = p
+	return b
+}
+
+func (b Builder) RichParameterValues(p []codersdk.WorkspaceBuildParameter) Builder {
+	// nolint: revive
+	b.richParameterValues = p
+	return b
+}
+
+// SetLastWorkspaceBuildInTx prepopulates the Builder's cache with the last workspace build.  This allows us
+// to avoid a repeated database query when the Builder's caller also needs the workspace build, e.g. auto-start &
+// auto-stop.
+//
+// CAUTION: only call this method from within a database transaction with RepeatableRead isolation.  This transaction
+// MUST be the database.Store you call Build() with.
+func (b Builder) SetLastWorkspaceBuildInTx(build *database.WorkspaceBuild) Builder {
+	// nolint: revive
+	b.lastBuild = build
+	return b
+}
+
+// SetLastWorkspaceBuildJobInTx prepopulates the Builder's cache with the last workspace build job.  This allows us
+// to avoid a repeated database query when the Builder's caller also needs the workspace build job, e.g. auto-start &
+// auto-stop.
+//
+// CAUTION: only call this method from within a database transaction with RepeatableRead isolation.  This transaction
+// MUST be the database.Store you call Build() with.
+func (b Builder) SetLastWorkspaceBuildJobInTx(job *database.ProvisionerJob) Builder {
+	// nolint: revive
+	b.lastBuildJob = job
+	return b
+}
+
+type BuildError struct {
+	// Status is a suitable HTTP status code
+	Status  int
+	Message string
+	Wrapped error
+}
+
+func (e BuildError) Error() string {
+	return e.Wrapped.Error()
+}
+
+func (e BuildError) Unwrap() error {
+	return e.Wrapped
+}
+
+// Build computes and inserts a new workspace build into the database.  If authFunc is provided, it also performs
+// authorization preflight checks.
+func (b *Builder) Build(
+	ctx context.Context,
+	store database.Store,
+	authFunc func(action rbac.Action, object rbac.Objecter) bool,
+) (
+	*database.WorkspaceBuild, *database.ProvisionerJob, error,
+) {
+	b.ctx = ctx
+
+	// Run the build in a transaction with RepeatableRead isolation, and retries.
+	// RepeatableRead isolation ensures that we get a consistent view of the database while
+	// computing the new build.  This simplifies the logic so that we do not need to worry if
+	// later reads are consistent with earlier ones.
+	var err error
+	for retries := 0; retries < 5; retries++ {
+		var workspaceBuild *database.WorkspaceBuild
+		var provisionerJob *database.ProvisionerJob
+		err := store.InTx(func(store database.Store) error {
+			b.store = store
```

Comment:
Do we need to be worried about a builder being used across multiple goroutines?

Reply:
Depends on what you mean.

When you are configuring the builder, it's passed around by value, so it's safe to use in multiple go routines.  However, it's not safe to call `b.Build()` from more than one goroutine.  It's hard to imagine why you would do that, as we explicitly ensure that only one build can be running at a time per workspace.

If, however, you're talking about multiple goroutines attempting to insert a build on the same logical workspace (i.e. the goroutines don't share memory, but they are operating on the same database object), then the use of database transactions will ensure at most one build is inserted.